File: QuickSort/quickSort.cpp
#include <bits/stdc++.h>
using namespace std;

int partitionArr(int *a,int s,int e){

    int pivot = a[e];

    int i=s-1;

    for(int j = s;j<e;j++){
        if(a[j]<pivot){
            i++;
            swap(a[j],a[i]);
        }
    }
    swap(a[e],a[i+1]);

    return i+1;
}



void quickSort(int a[],int s,int e){

    if(s>=e){
        return;
    }

    int c = partitionArr(a,s,e);

    quickSort(a,s,c-1);
    quickSort(a,c+1,e);

}



int main(){

    int n;
    cin>>n;

    int a[n];

    for(int i=0;i<n;i++){
        cin>>a[i];
    }

    quickSort(a,0,n-1);

    for(int i=0;i<n;i++){
        cout<<a[i]<<" ";
    }



    return 0;
}

File: quicksort/quick_sort.rs
use std::cmp::PartialOrd;

pub fn partition<T: PartialOrd>(arr: &mut [T], lo: isize, hi: isize) -> isize {
    let pivot = hi as usize;
    let mut i = lo - 1;
    let mut j = hi;

    loop {
        i += 1;
        while arr[i as usize] < arr[pivot] {
            i += 1;
        }
        j -= 1;
        while j >= 0 && arr[j as usize] > arr[pivot] {
            j -= 1;
        }
        if i >= j {
            break;
        } else {
            arr.swap(i as usize, j as usize);
        }
    }
    arr.swap(i as usize, pivot as usize);
    i
}

fn _quick_sort<T: Ord>(arr: &mut [T], lo: isize, hi: isize) {
    if lo < hi {
        let p = partition(arr, lo, hi);
        _quick_sort(arr, lo, p - 1);
        _quick_sort(arr, p + 1, hi);
    }
}

pub fn quick_sort<T: Ord>(arr: &mut [T]) {
    let len = arr.len();
    if len > 1 {
        _quick_sort(arr, 0, (len - 1) as isize);
    }
}

fn main() {  
    let mut array= [2, 1, 3, 5, 4];
    quick_sort(&mut array);
    assert_eq!(array, [1, 2, 3, 4, 5]);

    for i in 0..array.len() {
        println!("{} ", array[i]);
    }
}